shapeshift/web
# feat: thorchain savers withdraw sweep

## Description

The follow-up of https://github.com/shapeshift/web/pull/5614, for withdraws.

Additionally:

- improves perf so that react-escape-hatched queries use the same 5000ms `staleTime`, to avoid spamming THOR
- abstracts THOR `getHasEnoughXYZ` into reusable methods
- improves the minimum dummy 10_000 bps quote to handle missing `expected_amount_out` and `quote.fees.outbound` to ensure we get an outbound fees quote

## Pull Request Type


## Issue (if applicable)



closes https://github.com/shapeshift/web/issues/5619

## Risk



Withdraws are borked because of UTXO reconciliation and high fees currently, so this should only bring improvements.

## Testing



- Ensure withdraws trigger a sweep step when there aren't enough funds in your zeroth/active address
- Ensure withdraws do *not* trigger a sweep step after sending enough for withdraw dust amount + Tx  fees to your zeroth/active address
- Ensure EVM native/token assets as well as Cosmos withdraws are still happy

### Engineering



### Operations



## Screenshots (if applicable)

### Sweep needed

<img width="525" alt="Screenshot 2023-11-20 at 13 53 56" src="https://github.com/shapeshift/web/assets/17035424/05f7e781-61bc-4b8f-a0f0-67fc22e3c578">
<img width="525" alt="Screenshot 2023-11-20 at 13 53 56" src="https://github.com/shapeshift/web/assets/17035424/bdd978ea-8315-4d61-a700-f99acf660d22">
<img width="540" alt="Screenshot 2023-11-20 at 13 54 36" src="https://github.com/shapeshift/web/assets/17035424/a1c5c74f-a877-42c3-adfe-1ecdfafb7be2">
<img width="531" alt="Screenshot 2023-11-20 at 13 54 47" src="https://github.com/shapeshift/web/assets/17035424/3b15bf1f-f1cc-4ded-af00-b485de9e11da">
<img width="734" alt="image" src="https://github.com/shapeshift/web/assets/17035424/6261395d-fe42-4f2f-a630-14c22f44f995">

### Sweep not needed i.e all funds in zeroth address

<img width="537" alt="image" src="https://github.com/shapeshift/web/assets/17035424/a3e68935-5b71-45a9-b88d-29cb6daca0e3">
<img width="527" alt="image" src="https://github.com/shapeshift/web/assets/17035424/fba4b8c8-1a0e-4df4-bf80-929203ded4d3">
<img width="518" alt="image" src="https://github.com/shapeshift/web/assets/17035424/5b60e30d-5dd7-42e2-9407-f68cc4490a65">
<img width="621" alt="image" src="https://github.com/shapeshift/web/assets/17035424/1f5637db-37dc-490e-89bb-6803995055a6">
 
### Minimum not met

<img width="534" alt="image" src="https://github.com/shapeshift/web/assets/17035424/a1d2a27d-096d-4b64-90f0-ad546a472a72">

### Minimum met, but too low fees (UX to be confirmed by @reallybeard)

<img width="529" alt="image" src="https://github.com/shapeshift/web/assets/17035424/09b822fd-787e-48d9-b4a5-3dc97ccca6f9">

### EVM withdraw

<img width="544" alt="image" src="https://github.com/shapeshift/web/assets/17035424/1cf97689-0a48-443e-be9e-7b9088c326b8">
<img width="535" alt="image" src="https://github.com/shapeshift/web/assets/17035424/fa37b843-d9c5-4081-b664-415997ae4f8a">
<img width="538" alt="image" src="https://github.com/shapeshift/web/assets/17035424/7bf07f61-2056-4c86-b17f-6bddf1daaed0">
<img width="729" alt="image" src="https://github.com/shapeshift/web/assets/17035424/a59e9965-be81-4d4e-8f96-0e77c58e3f21">

### ATOM withdraw

<img width="551" alt="image" src="https://github.com/shapeshift/web/assets/17035424/ba928087-7351-4c7b-8d9c-524a32e2deb2">
<img width="540" alt="image" src="https://github.com/shapeshift/web/assets/17035424/8eff77ba-adcb-423f-95d3-5c176f668da5">
<img width="529" alt="image" src="https://github.com/shapeshift/web/assets/17035424/4e1a1e52-034f-4d25-a719-7240ca027de1">
<img width="733" alt="image" src="https://github.com/shapeshift/web/assets/17035424/008a58e3-19dc-4a3d-bbce-5ebce9a51958">
